File: android-app/src/main/java/com/prevengos/plug/android/ui/CaptureViewModel.java
package com.prevengos.plug.android.ui;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.prevengos.plug.android.PrevengosApplication;
import com.prevengos.plug.android.data.local.entity.PacienteEntity;
import com.prevengos.plug.android.data.local.entity.RespuestaLocal;
import com.prevengos.plug.android.data.repository.CuestionarioRepository;
import com.prevengos.plug.android.data.repository.PacienteRepository;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class CaptureViewModel extends ViewModel {
    private static final String DEFAULT_PLANTILLA = "anamnesis_inicial";

    private final PacienteRepository pacienteRepository;
    private final CuestionarioRepository cuestionarioRepository;
    private final PrevengosApplication application;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    private final MutableLiveData<CaptureUiState> uiState = new MutableLiveData<>(new CaptureUiState());
    private final LiveData<String> pacientes;

    public CaptureViewModel(
            PacienteRepository pacienteRepository,
            CuestionarioRepository cuestionarioRepository,
            PrevengosApplication application
    ) {
        this.pacienteRepository = pacienteRepository;
        this.cuestionarioRepository = cuestionarioRepository;
        this.application = application;
        this.pacientes = Transformations.map(
                pacienteRepository.observePacientes(),
                this::buildResumen
        );
    }

    public LiveData<CaptureUiState> getUiState() {
        return uiState;
    }

    public LiveData<String> getPacientes() {
        return pacientes;
    }

    public void guardarCaptura(String nif, String nombre, String apellidos, String telefono, String email, String nota) {
        if (nif == null || nif.trim().isEmpty() || nombre == null || nombre.trim().isEmpty() || apellidos == null || apellidos.trim().isEmpty()) {
            uiState.postValue(new CaptureUiState(false, null, "Los campos NIF, nombre y apellidos son obligatorios"));
            return;
        }
        uiState.postValue(new CaptureUiState(true, null, null));
        executor.execute(() -> {
            try {
                String telefonoValue = telefono != null && !telefono.trim().isEmpty() ? telefono : null;
                String emailValue = email != null && !email.trim().isEmpty() ? email : null;
                String notaValue = nota != null && !nota.trim().isEmpty() ? nota : null;
                PacienteEntity paciente = pacienteRepository.createPaciente(nif.trim(), nombre.trim(), apellidos.trim(), telefonoValue, emailValue);
                RespuestaLocal respuesta = new RespuestaLocal("nota_inicial", notaValue, null, null);
                cuestionarioRepository.createDraft(
                        paciente.getPacienteId(),
                        DEFAULT_PLANTILLA,
                        Collections.singletonList(respuesta)
                );
                application.triggerOneTimeSync();
                uiState.postValue(new CaptureUiState(false, "Captura guardada para " + paciente.getNombre(), null));
            } catch (Exception exception) {
                String errorMessage = exception.getMessage() != null ? exception.getMessage() : "Error desconocido";
                uiState.postValue(new CaptureUiState(false, null, errorMessage));
            }
        });
    }

    public void acknowledgeFeedback() {
        uiState.postValue(new CaptureUiState());
    }

    private String buildResumen(List<PacienteEntity> pacientes) {
        if (pacientes == null || pacientes.isEmpty()) {
            return "";
        }
        StringBuilder builder = new StringBuilder();
        for (PacienteEntity paciente : pacientes) {
            if (builder.length() > 0) {
                builder.append('\n');
            }
            builder.append(paciente.getNombre()).append(' ').append(paciente.getApellidos());
        }
        return builder.toString();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        executor.shutdownNow();
    }
}

File: android-app/src/main/java/com/prevengos/plug/android/ui/CaptureViewModel.kt
package com.prevengos.plug.android.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.prevengos.plug.android.PrevengosApplication
import com.prevengos.plug.android.data.local.entity.RespuestaLocal
import com.prevengos.plug.android.data.repository.CuestionarioRepository
import com.prevengos.plug.android.data.repository.PacienteRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class CaptureViewModel(
    private val pacienteRepository: PacienteRepository,
    private val cuestionarioRepository: CuestionarioRepository,
    private val application: PrevengosApplication
) : ViewModel() {

    private val _uiState = MutableStateFlow(CaptureUiState())
    val uiState: StateFlow<CaptureUiState> = _uiState

    val pacientes = pacienteRepository.observePacientes()
        .map { pacientes -> pacientes.joinToString(separator = "\n") { "${it.nombre} ${it.apellidos}" } }
        .catch { emit("") }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), "")

    fun guardarCaptura(
        nif: String,
        nombre: String,
        apellidos: String,
        telefono: String?,
        email: String?,
        nota: String
    ) {
        viewModelScope.launch {
            if (nif.isBlank() || nombre.isBlank() || apellidos.isBlank()) {
                _uiState.value = CaptureUiState(error = "Los campos NIF, nombre y apellidos son obligatorios")
                return@launch
            }
            _uiState.value = CaptureUiState(isSaving = true)
            try {
                val paciente = pacienteRepository.createPaciente(
                    nif = nif,
                    nombre = nombre,
                    apellidos = apellidos,
                    telefono = telefono?.takeIf { it.isNotBlank() },
                    email = email?.takeIf { it.isNotBlank() }
                )
                cuestionarioRepository.createDraft(
                    pacienteId = paciente.pacienteId,
                    plantillaCodigo = DEFAULT_PLANTILLA,
                    respuestas = listOf(
                        RespuestaLocal(
                            preguntaCodigo = "nota_inicial",
                            valor = nota.ifBlank { null }
                        )
                    )
                )
                application.triggerOneTimeSync()
                _uiState.value = CaptureUiState(message = "Captura guardada para ${paciente.nombre}")
            } catch (exception: Exception) {
                _uiState.value = CaptureUiState(error = exception.message ?: "Error desconocido")
            }
        }
    }

    fun acknowledgeFeedback() {
        _uiState.value = CaptureUiState()
    }

    companion object {
        private const val DEFAULT_PLANTILLA = "anamnesis_inicial"
    }
}
